## Homework

I came across my kid's homework which was some kind of simulation
to calculate the probability that a random walk of length 10 didn't leave $[-2,2]$. 

The code was so ugly, an affront to [the Zen of Python](https://www.python.org/dev/peps/pep-0020/),
that I immediately scribbled all over it in red ink.


This is my final code

In [343]:
%%time
import numpy as np

batch_size = 10**6
walk_length = 10
#make a random matrix and center the values
M = np.random.randint(3, size=(batch_size, walk_length))
M -= np.ones_like(M)

#make an upper triangular matrix and calculate postions
UT = np.triu(np.ones((10,10)))
positions = np.dot(M,UT)

#calculate fancy indexing and use fancy indexing to finish
max_dist_to_0  = np.abs(positions).max(axis=1)
print len(max_dist_to_0[ max_dist_to_0 < 2])

137835
CPU times: user 447 ms, sys: 189 ms, total: 635 ms
Wall time: 722 ms


###  My first version

- tidied up the while loop
- use and in statement instead of repeated comparisons
- the += operator here and there

For me this still looks like someone that does Java/C++.

In [335]:
%%time
from random import randint

def une_marche():
    x = y = 0
    while  -1 <= x <= 1 and y < 10:
        a = randint(-1,1)
        x += a
        y += 1
    if x in [-2,2]:
        return 0
    else:
        return 1
    
T = 0
num_tries = 10**6
for k in xrange(num_tries):
    T += une_marche()
print float(T)/num_tries    

0.137696
CPU times: user 17.6 s, sys: 90.3 ms, total: 17.7 s
Wall time: 17.7 s


## Second version

This is somewhat simpler and uses standard Python idioms

- for break else  statement
- list comprehension [ x for x in list]
- the += operator

It runs marginally faster than the original


In [334]:
%%time
def une_marche2():
    x = 0
    for y in range(10): 
        x += randint(-1,1)
        if not( -1 <= x <= 1): break
    else:
        return 1
    return 0
  
num_tries = 10**5
tries = [une_marche2() for k in xrange(num_tries)]
    
print float(sum(tries))/num_tries

0.13756
CPU times: user 1.73 s, sys: 26.9 ms, total: 1.76 s
Wall time: 1.75 s


## The final version

This is basically about vectorisation and matrix multiplication.

Just dotting with a vector of ones doesn't work
because the walk can break out of $[-1,1]$ and get back in again.
So we use an upper triangular matrix to calculate a running total 
of the position deltas.

- I don't really care about memory usage
- this is definitely not for les tapettes
- finally optimisation of counting solutions is tricky

### Optimisation

Finally I found this which is kind of obvious as an afterthought
https://stackoverflow.com/questions/36267936/normalizing-rows-of-a-matrix-python


CPU times: user 411 ms, sys: 84.4 ms, total: 495 ms

This compares very favorably with my first version
running on 10**6 :

CPU times: user 17.6 s

In [321]:
%%time
import numpy as np

n = 10**6
M = np.random.randint(3, size=(n,10 ))
M -= np.ones_like(M)

Y = np.triu(np.ones((10,10)))
walks = np.dot(M,Y)

#this is a bit fecked but I don't see how 
#to easily vectorize this
#CPU times: user 3.71 s, sys: 229 ms, total: 3.94 s
#bounds = np.array([max(abs(x)) for x in walks])
#len(bounds[bounds < 2])

#this didn't work - twise as long on 10**6
#CPU times: user 440 ms, sys: 86.5 ms, total: 527 ms
lengths = np.abs(walks).max(axis=1)
print len(lengths[lengths < 2])



137831
CPU times: user 411 ms, sys: 84.4 ms, total: 495 ms
Wall time: 384 ms
